lib/digital_public_works_web/controllers/project_invite_controller.ex

```diff
@@ -0,0 +1,153 @@
+  plug :check_auth
+
+  defp get_project(%{params: %{"project_id" => id}} = conn, _args) do
+    conn |> assign(:project, Projects.get_project!(id))
+  end
+
+  defp get_project(conn, _args), do: conn
+
+  defp get_project_invite(%{params: %{"id" => id}} = conn, _args) do
+    try do
+      project_invite = Accounts.get_project_invite!(id)
+
+      project_invite =
+        if project = conn.assigns[:project] do
+          Map.put(project_invite, :project, project)
+        else
+          project_invite |> Repo.preload(:project)
+        end
+
+      conn |> assign(:project_invite, project_invite)
+    rescue
+      _ ->
+        conn
+        |> put_flash(:error, "Invite not found.")
+        |> redirect(to: Routes.page_path(conn, :index))
+        |> halt()
+    end
+  end
+
+  defp get_project_invite(conn, _args) do
+    project = conn.assigns.project
+
+    project_invite = Ecto.build_assoc(project, :project_invites, project: project)
+
+    conn |> assign(:project_invite, project_invite)
+  end
+
+  defp check_auth(conn, _args) do
+    cond do
+      can?(conn.assigns.current_user, action_name(conn), conn.assigns.project_invite) ->
+        conn
+
+      conn.assigns.current_user ->
+        conn
+        |> put_flash(:error, "You don't have access to that")
+        |> redirect(to: Routes.page_path(conn, :index))
+        |> halt()
+
+      true ->
+        conn
+        |> put_flash(:info, "You need to log in first")
+        |> redirect(to: Routes.session_path(conn, :new))
+        |> halt()
+    end
+  end
+
+  def index(conn, _params) do
+    project = conn.assigns.project
+
+    conn
+    |> assign(:project_invites, project |> project_invites)
+    |> assign(:users, project |> project_users)
+    |> assign(:changeset, Accounts.change_project_invite())
+    |> render("index.html")
+  end
+
+  def create(conn, %{"project_invite" => params}) do
+    project = conn.assigns.project
+
+    case Accounts.create_project_invite(conn.assigns.project, params) do
+      {:ok, project_invite} ->
+        Email.project_invite_email(project_invite |> Map.put(:project, project)) |> Mailer.deliver_now
+
+        conn
+        |> put_flash(:info, "Invite sent.")
+        |> redirect(to: Routes.project_invite_path(conn, :index, project))
+
+      {:error, %Ecto.Changeset{} = changeset} ->
+        conn
+        |> assign(:project_invites, project |> project_invites)
+        |> assign(:users, project |> project_users)
+        |> assign(:changeset, changeset)
+        |> render("index.html")
+    end
+  end
+
+  def update(conn, _params) do
+    user = conn.assigns.current_user
+    invite = conn.assigns.project_invite
+
+    case Accounts.update_project_invite(invite, user) do
+      {:ok, _project_invite} ->
+        conn
+        |> put_flash(:info, "Joined project.")
+        |> redirect(to: Routes.project_path(conn, :show, invite.project))
+      {:error, changeset} ->
+        error_message =
+          changeset.errors
```

Comment:
Not a big change, but https://hexdocs.pm/ecto/Ecto.Changeset.html#traverse_errors/2 will also find nested errors in the event they exist. So if you ever end up making a helper for this, you'll likely want traverse_errors